File: app/src/main/java/com/example/mainactivity/ui/shopping/ShoppingScreens.kt
@file:Suppress("ktlint:standard:function-naming")

package com.example.mainactivity.ui.shopping

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.Send
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.ChevronRight
import androidx.compose.material.icons.filled.MoreVert
import androidx.compose.material.icons.filled.ShoppingCart
import androidx.compose.material.icons.outlined.Circle
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExtendedFloatingActionButton
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.focus.onFocusChanged
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import com.example.mainactivity.data.ShoppingItemModel
import com.example.mainactivity.ui.components.EmptyState
import com.example.mainactivity.ui.components.FeatureTopBar
import com.example.mainactivity.ui.components.InputDialog
import com.example.mainactivity.ui.components.LoadingState
import com.example.mainactivity.ui.components.PillTag
import com.example.mainactivity.ui.components.RefreshOnResume
import com.example.mainactivity.ui.components.SwipeToRevealDelete

@Composable
fun ShoppingScreen(
    onBack: () -> Unit,
    onOpenList: (String) -> Unit,
    viewModel: ShoppingViewModel = viewModel(),
) {
    val lists by viewModel.lists.collectAsStateWithLifecycle(emptyList())
    val isLoading by viewModel.isLoading.collectAsStateWithLifecycle(false)
    var showAdd by remember { mutableStateOf(false) }

    RefreshOnResume { viewModel.refresh() }

    Scaffold(
        containerColor = MaterialTheme.colorScheme.background,
        topBar = { FeatureTopBar("Shopping lists", onBack) },
        floatingActionButton = {
            ExtendedFloatingActionButton(
                onClick = { showAdd = true },
                icon = { Icon(Icons.Filled.Add, null) },
                text = { Text("New list") },
                containerColor = MaterialTheme.colorScheme.primary,
                contentColor = MaterialTheme.colorScheme.onPrimary,
            )
        },
    ) { padding ->
        if (isLoading) {
            Box(Modifier.fillMaxSize().padding(padding), contentAlignment = Alignment.Center) {
                LoadingState()
            }
        } else if (lists.isEmpty()) {
            Box(Modifier.fillMaxSize().padding(padding), contentAlignment = Alignment.Center) {
                EmptyState(Icons.Filled.ShoppingCart, "No lists yet", "Create a shared shopping list for your family.")
            }
        } else {
            LazyColumn(
                Modifier.fillMaxSize().padding(padding),
                contentPadding = PaddingValues(20.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                items(lists, key = { it.id }) { list ->
                    SwipeToRevealDelete(onDelete = { viewModel.deleteList(list) }, shape = RoundedCornerShape(20.dp)) {
                        Surface(
                            onClick = { onOpenList(list.id) },
                            shape = RoundedCornerShape(20.dp),
                            color = MaterialTheme.colorScheme.surface,
                            shadowElevation = 2.dp,
                            modifier = Modifier.fillMaxWidth(),
                        ) {
                            Row(Modifier.padding(18.dp), verticalAlignment = Alignment.CenterVertically) {
                                Box(
                                    Modifier.size(44.dp),
                                    contentAlignment = Alignment.Center,
                                ) {
                                    Icon(Icons.Filled.ShoppingCart, null, tint = MaterialTheme.colorScheme.primary)
                                }
                                Spacer(Modifier.size(8.dp))
                                Text(list.title, Modifier.weight(1f), style = MaterialTheme.typography.titleMedium, color = MaterialTheme.colorScheme.onSurface)
                                Icon(Icons.Filled.ChevronRight, null, tint = MaterialTheme.colorScheme.onSurfaceVariant)
                            }
                        }
                    }
                }
            }
        }
    }

    if (showAdd) {
        InputDialog(
            title = "New shopping list",
            label = "List name",
            onDismiss = { showAdd = false },
            onConfirm = { value, _ ->
                viewModel.addList(value)
                showAdd = false
            },
        )
    }
}

@Composable
fun ShoppingDetailScreen(
    listId: String,
    onBack: () -> Unit,
    viewModel: ShoppingViewModel = viewModel(),
) {
    androidx.compose.runtime.LaunchedEffect(listId) { viewModel.loadListDetail(listId) }
    val list by viewModel.selectedList.collectAsStateWithLifecycle()
    val items by viewModel.items.collectAsStateWithLifecycle()
    var newItemText by remember { mutableStateOf("") }
    var showMenu by remember { mutableStateOf(false) }
    var showRename by remember { mutableStateOf(false) }
    val remaining = items.count { !it.checked }

    fun addItem() {
        val text = newItemText.trim()
        if (text.isNotBlank()) {
            viewModel.addItem(listId, text)
            newItemText = ""
        }
    }

    Scaffold(
        modifier = Modifier.imePadding(),
        containerColor = MaterialTheme.colorScheme.background,
        topBar = {
            FeatureTopBar(list?.title ?: "List", onBack) {
                if (items.isNotEmpty()) {
                    PillTag(
                        "$remaining left",
                        MaterialTheme.colorScheme.primaryContainer,
                        MaterialTheme.colorScheme.onPrimaryContainer,
                        Modifier.padding(end = 4.dp),
                    )
                }
                Box {
                    IconButton(onClick = { showMenu = true }) {
                        Icon(Icons.Filled.MoreVert, contentDescription = "More options")
                    }
                    DropdownMenu(
                        expanded = showMenu,
                        onDismissRequest = { showMenu = false },
                    ) {
                        DropdownMenuItem(
                            text = { Text("Rename list") },
                            onClick = {
                                showMenu = false
                                showRename = true
                            },
                        )
                    }
                }
            }
        },
        bottomBar = {
            Surface(
                color = MaterialTheme.colorScheme.surface,
                shadowElevation = 8.dp,
                modifier = Modifier.fillMaxWidth(),
            ) {
                Row(
                    modifier = Modifier.padding(horizontal = 12.dp, vertical = 8.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    OutlinedTextField(
                        value = newItemText,
                        onValueChange = { newItemText = it },
                        placeholder = { Text("Add item…") },
                        singleLine = true,
                        shape = RoundedCornerShape(24.dp),
                        keyboardOptions = KeyboardOptions(imeAction = ImeAction.Done),
                        keyboardActions = KeyboardActions(onDone = { addItem() }),
                        modifier = Modifier.weight(1f),
                    )
                    Spacer(Modifier.width(8.dp))
                    IconButton(
                        onClick = { addItem() },
                        enabled = newItemText.isNotBlank(),
                    ) {
                        Icon(
                            Icons.AutoMirrored.Filled.Send,
                            contentDescription = "Add item",
                            tint = if (newItemText.isNotBlank()) MaterialTheme.colorScheme.primary else MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                }
            }
        },
    ) { padding ->
        if (items.isEmpty()) {
            Box(Modifier.fillMaxSize().padding(padding), contentAlignment = Alignment.Center) {
                EmptyState(Icons.Filled.ShoppingCart, "Empty list", "Tap the field below to add your first item.")
            }
        } else {
            LazyColumn(
                Modifier.fillMaxSize().padding(padding),
                contentPadding = PaddingValues(20.dp),
                verticalArrangement = Arrangement.spacedBy(10.dp),
            ) {
                items(items, key = { it.id }) { item -> ShoppingItemRow(item, viewModel) }
            }
        }
    }

    if (showRename) {
        InputDialog(
            title = "Rename list",
            label = "List name",
            initial = list?.title ?: "",
            onDismiss = { showRename = false },
            onConfirm = { value, _ ->
                viewModel.renameList(listId, value)
                showRename = false
            },
        )
    }
}

@Composable
private fun ShoppingItemRow(
    item: ShoppingItemModel,
    viewModel: ShoppingViewModel,
) {
    var isEditing by remember { mutableStateOf(false) }
    var editText by remember { mutableStateOf(item.item) }
    val focusRequester = remember { FocusRequester() }

    fun commitEdit() {
        if (!isEditing) return
        val trimmed = editText.trim()
        isEditing = false
        if (trimmed.isNotBlank() && trimmed != item.item) {
            viewModel.renameItem(item, trimmed)
        } else {
            editText = item.item
        }
    }

    LaunchedEffect(isEditing) {
        if (isEditing) focusRequester.requestFocus()
    }

    SwipeToRevealDelete(onDelete = { viewModel.deleteItem(item) }, shape = RoundedCornerShape(16.dp)) {
        Surface(
            shape = RoundedCornerShape(16.dp),
            color = MaterialTheme.colorScheme.surface,
            modifier = Modifier.fillMaxWidth(),
        ) {
            Row(Modifier.padding(horizontal = 12.dp, vertical = 10.dp), verticalAlignment = Alignment.CenterVertically) {
                IconButton(onClick = { viewModel.toggle(item) }) {
                    Icon(
                        if (item.checked) Icons.Filled.CheckCircle else Icons.Outlined.Circle,
                        null,
                        tint = if (item.checked) MaterialTheme.colorScheme.primary else MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
                if (isEditing) {
                    BasicTextField(
                        value = editText,
                        onValueChange = { editText = it },
                        textStyle = MaterialTheme.typography.bodyLarge.copy(
                            color = MaterialTheme.colorScheme.onSurface,
                        ),
                        cursorBrush = SolidColor(MaterialTheme.colorScheme.primary),
                        singleLine = true,
                        keyboardOptions = KeyboardOptions(imeAction = ImeAction.Done),
                        keyboardActions = KeyboardActions(onDone = { commitEdit() }),
                        modifier = Modifier
                            .weight(1f)
                            .focusRequester(focusRequester)
                            .onFocusChanged { if (!it.isFocused) commitEdit() },
                    )
                } else {
                    Text(
                        item.item,
                        modifier = Modifier
                            .weight(1f)
                            .clickable { isEditing = true },
                        style = MaterialTheme.typography.bodyLarge,
                        color = if (item.checked) MaterialTheme.colorScheme.onSurfaceVariant else MaterialTheme.colorScheme.onSurface,
                        textDecoration = if (item.checked) TextDecoration.LineThrough else TextDecoration.None,
                    )
                }
            }
        }
    }
}

File: ios/FamilyApp/Features/Shopping/ShoppingScreens.swift
// Shopping screens — the iOS twin of ShoppingScreens.kt: list overview with progress
// labels + swipe-to-delete, detail with active/completed sections, inline rename,
// bottom add-item bar, rename/change-icon menu.
import SwiftUI

struct ShoppingScreen: View {
    let viewModel: ShoppingViewModel
    let onOpenList: (String) -> Void

    @State private var showAdd = false

    var body: some View {
        ZStack(alignment: .bottomTrailing) {
            Group {
                if viewModel.isLoading {
                    LoadingState()
                        .frame(maxHeight: .infinity, alignment: .top)
                } else if viewModel.lists.isEmpty {
                    EmptyState(
                        systemImage: "cart.fill",
                        title: L("No lists yet"),
                        subtitle: L("Create a shared shopping list for your family.")
                    )
                } else {
                    List {
                        ForEach(viewModel.lists) { list in
                            ShoppingListRow(
                                list: list,
                                progress: viewModel.listProgress[list.id]
                            ) { onOpenList(list.id) }
                                .listRowSeparator(.hidden)
                                .listRowBackground(Color.clear)
                                .listRowInsets(EdgeInsets(
                                    top: 6, leading: Spacing.screenEdge,
                                    bottom: 6, trailing: Spacing.screenEdge
                                ))
                                .swipeActions(edge: .trailing) {
                                    Button(role: .destructive) {
                                        viewModel.deleteList(list)
                                    } label: {
                                        Label("Delete", systemImage: "trash")
                                    }
                                }
                        }
                    }
                    .listStyle(.plain)
                    .scrollContentBackground(.hidden)
                    .refreshable { viewModel.refresh() }
                }
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity)

            FloatingActionButton(text: L("New list"), systemImage: "plus") { showAdd = true }
        }
        .ambientBackground()
        .featureTopBar(L("Shopping lists"))
        .resumeEffect { viewModel.refresh() }
        .sheet(isPresented: $showAdd) {
            NewListSheet { title, icon, color in
                viewModel.addList(title: title, icon: icon, color: color)
                showAdd = false
            }
        }
    }
}

private struct ShoppingListRow: View {
    let list: ShoppingListModel
    let progress: ListProgress?
    let onTap: () -> Void

    private var fraction: Double {
        guard let progress, progress.total > 0 else { return 0 }
        return Double(progress.bought) / Double(progress.total)
    }

    private var allDone: Bool {
        guard let progress else { return false }
        return progress.total > 0 && progress.bought == progress.total
    }

    var body: some View {
        Button(action: onTap) {
            VStack(spacing: 12) {
                HStack(spacing: 12) {
                    if allDone {
                        RoundedRectangle(cornerRadius: Radius.badgeLarge, style: .continuous)
                            .fill(Color.appSuccess.opacity(0.13))
                            .frame(width: 44, height: 44)
                            .overlay(Image(systemName: "checkmark")
                                .font(.system(size: 19, weight: .bold))
                                .foregroundStyle(Color.appSuccess))
                    } else {
                        FeatureBadge(
                            systemImage: IconKeyMap.shoppingSymbol(list.icon),
                            feature: .shopping,
                            size: 44,
                            cornerRadius: Radius.badgeLarge,
                            colorOverride: hexColor(list.color)
                        )
                    }
                    VStack(alignment: .leading, spacing: 2) {
                        Text(list.title)
                            .font(.system(size: 15.5, weight: .semibold))
                            .foregroundStyle(allDone ? Color.appOnSurfaceVariant : Color.appOnSurface)
                        Text(shoppingProgressLabel(progress, locale: appLocale))
                            .font(.caption)
                            .foregroundStyle(allDone ? Color.appSuccess : Color.appCaption)
                    }
                    Spacer()
                    Image(systemName: "chevron.right")
                        .font(.system(size: 13, weight: .semibold))
                        .foregroundStyle(Color.appCaption)
                }
                if !allDone, progress != nil {
                    GlassProgressBar(value: fraction, tint: .appPrimary, height: 4)
                }
            }
            .padding(Spacing.cardPadding)
            .modifier(ListRowSurface(ghost: allDone))
        }
        .buttonStyle(PressScaleButtonStyle())
    }
}

/// Glass for active surfaces, dashed ghost for completed ones (spec 2d).
private struct ListRowSurface: ViewModifier {
    let ghost: Bool
    var cornerRadius: CGFloat = Radius.overviewCard
    func body(content: Content) -> some View {
        if ghost {
            content.ghostSurface(cornerRadius: cornerRadius)
        } else {
            content.glassCard(cornerRadius: cornerRadius)
        }
    }
}

// MARK: - Detail

struct ShoppingDetailScreen: View {
    let listId: String
    let viewModel: ShoppingViewModel

    @State private var newItemText = ""
    @State private var showRename = false
    @State private var showChangeIcon = false
    @State private var showCompleted = true
    @State private var renameText = ""

    private var active: [ShoppingItemModel] {
        viewModel.items.filter { !$0.checked }
    }

    private var completed: [ShoppingItemModel] {
        viewModel.items.filter(\.checked)
    }

    private var remaining: Int {
        active.count
    }

    var body: some View {
        VStack(spacing: 0) {
            if viewModel.items.isEmpty {
                EmptyState(
                    systemImage: "cart.fill",
                    title: L("Empty list"),
                    subtitle: L("Tap the field below to add your first item.")
                )
                .frame(maxHeight: .infinity)
            } else {
                List {
                    ForEach(active) { item in
                        ShoppingItemRow(item: item, viewModel: viewModel)
                            .shoppingRowStyle()
                            .swipeActions(edge: .trailing) {
                                Button(role: .destructive) {
                                    viewModel.deleteItem(item)
                                } label: {
                                    Label("Delete", systemImage: "trash")
                                }
                            }
                    }
                    if !completed.isEmpty {
                        CompletedHeader(count: completed.count, expanded: showCompleted) {
                            withAnimation { showCompleted.toggle() }
                        }
                        .listRowSeparator(.hidden)
                        .listRowBackground(Color.clear)
                        if showCompleted {
                            ForEach(completed) { item in
                                ShoppingItemRow(item: item, viewModel: viewModel)
                                    .shoppingRowStyle()
                                    .swipeActions(edge: .trailing) {
                                        Button(role: .destructive) {
                                            viewModel.deleteItem(item)
                                        } label: {
                                            Label("Delete", systemImage: "trash")
                                        }
                                    }
                            }
                        }
                    }
                }
                .listStyle(.plain)
                .scrollContentBackground(.hidden)
            }

            addItemBar
        }
        .ambientBackground()
        .featureTopBar(viewModel.selectedList?.title ?? L("List"))
        .toolbar {
            ToolbarItemGroup(placement: .topBarTrailing) {
                if !viewModel.items.isEmpty {
                    Text("\(remaining) left")
                        .font(.system(size: 12, weight: .semibold))
                        .padding(.horizontal, 10)
                        .padding(.vertical, 4)
                        .background(Color.appPrimary.opacity(0.12), in: Capsule())
                        .foregroundStyle(Color.appPrimary)
                        .padding(.leading, 8)
                }
                Menu {
                    Button {
                        renameText = viewModel.selectedList?.title ?? ""
                        showRename = true
                    } label: {
                        Label(L("Rename list"), systemImage: "pencil")
                    }
                    Button { showChangeIcon = true } label: {
                        Label(L("Change icon"), systemImage: "star")
                    }
                    if !completed.isEmpty {
                        Button(role: .destructive) {
                            viewModel.clearCompleted(listId: listId)
                        } label: {
                            Label("\(L("Clear completed")) (\(completed.count))", systemImage: "trash")
                        }
                    }
                } label: {
                    Image(systemName: "ellipsis.circle")
                        .accessibilityLabel("More options")
                }
            }
        }
        .task(id: listId) { viewModel.loadListDetail(listId) }
        .inputDialog(
            isPresented: $showRename,
            title: L("Rename list"),
            label: L("List name"),
            text: $renameText
        ) {
            if !renameText.trimmingCharacters(in: .whitespaces).isEmpty {
                viewModel.renameList(listId: listId, newTitle: renameText.trimmingCharacters(in: .whitespaces))
            }
        }
        .sheet(isPresented: $showChangeIcon) {
            IconPickerSheet(
                title: L("Change icon"),
                options: IconOptions.shopping,
                selected: viewModel.selectedList?.icon ?? "shopping_cart",
                symbolFor: IconKeyMap.shoppingSymbol,
                onPick: { icon in
                    viewModel.changeListIcon(listId: listId, icon: icon)
                    showChangeIcon = false
                },
                initialColor: viewModel.selectedList?.color,
                onColorPick: { color in viewModel.changeListColor(listId: listId, color: color) }
            )
        }
    }

    private var addItemBar: some View {
        HStack(spacing: 10) {
            TextField("Add item…", text: $newItemText)
                .font(.system(size: 16))
                .tint(Color.appPrimary)
                .padding(.horizontal, Spacing.lg)
                .frame(height: 46)
                .glassCard(cornerRadius: 23)
                .onSubmit(addItem)
            Button(action: addItem) {
                Image(systemName: "arrow.up")
                    .font(.system(size: 18, weight: .bold))
                    .foregroundStyle(.white)
                    .frame(width: 46, height: 46)
                    .background(Color.appPrimary.opacity(newItemText.isEmpty ? 0.4 : 1), in: Circle())
                    .accentGlow(active: !newItemText.isEmpty, opacity: 0.35, radius: 8, y: 3)
            }
            .disabled(newItemText.isEmpty)
            .accessibilityLabel("Add item")
        }
        .padding(.horizontal, Spacing.md)
        .padding(.vertical, Spacing.sm)
        .background(.ultraThinMaterial)
    }

    private func addItem() {
        let text = newItemText.trimmingCharacters(in: .whitespaces)
        guard !text.isEmpty else { return }
        viewModel.addItem(listId: listId, item: text)
        newItemText = ""
    }
}

extension View {
    fileprivate func shoppingRowStyle() -> some View {
        listRowSeparator(.hidden)
            .listRowBackground(Color.clear)
            .listRowInsets(EdgeInsets(
                top: 5, leading: Spacing.screenEdge, bottom: 5, trailing: Spacing.screenEdge
            ))
    }
}

private struct ShoppingItemRow: View {
    let item: ShoppingItemModel
    let viewModel: ShoppingViewModel

    @State private var isEditing = false
    @State private var editText = ""
    @FocusState private var editFocused: Bool

    var body: some View {
        HStack(spacing: Spacing.md) {
            Button {
                UIImpactFeedbackGenerator(style: .light).impactOccurred()
                viewModel.toggle(item)
            } label: {
                if item.checked {
                    Circle().fill(Color.appPrimary)
                        .frame(width: 24, height: 24)
                        .overlay(Image(systemName: "checkmark")
                            .font(.system(size: 12, weight: .bold)).foregroundStyle(.white))
                } else {
                    Circle().strokeBorder(Color(hex: 0x9CA2BC), lineWidth: 1.8)
                        .frame(width: 24, height: 24)
                }
            }
            .buttonStyle(.plain)

            if isEditing {
                TextField("Item", text: $editText)
                    .font(.system(size: 15.5))
                    .tint(Color.appPrimary)
                    .focused($editFocused)
                    .onSubmit(commitEdit)
                    .onChange(of: editFocused) { _, focused in
                        if !focused { commitEdit() }
                    }
            } else {
                Text(item.item)
                    .font(.system(size: 15.5))
                    .foregroundStyle(item.checked ? Color(hex: 0xA6ACC4) : Color.appOnSurface)
                    .strikethrough(item.checked)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .contentShape(Rectangle())
                    .onTapGesture {
                        editText = item.item
                        isEditing = true
                        editFocused = true
                    }
            }
        }
        .padding(.horizontal, Spacing.lg)
        .padding(.vertical, 12)
        .modifier(ListRowSurface(ghost: item.checked, cornerRadius: Radius.row))
    }

    private func commitEdit() {
        guard isEditing else { return }
        isEditing = false
        let trimmed = editText.trimmingCharacters(in: .whitespaces)
        if !trimmed.isEmpty, trimmed != item.item {
            viewModel.renameItem(item, newName: trimmed)
        }
    }
}

private struct CompletedHeader: View {
    let count: Int
    let expanded: Bool
    let onToggle: () -> Void

    var body: some View {
        Button(action: onToggle) {
            HStack {
                Text(L("Completed (\(count))").uppercased())
                    .font(.sectionLabel)
                    .tracking(0.6)
                    .foregroundStyle(Color.appCaption)
                Spacer()
                Image(systemName: expanded ? "chevron.up" : "chevron.down")
                    .font(.system(size: 12, weight: .semibold))
                    .foregroundStyle(Color.appCaption)
            }
            .padding(.horizontal, Spacing.sm)
            .padding(.vertical, 6)
        }
        .buttonStyle(.plain)
        .accessibilityLabel(expanded ? L("Hide completed items") : L("Show completed items"))
    }
}

// MARK: - Shared pieces (also used by meals)

/// Extended floating action button — Liquid Glass FAB (solid accent capsule + glow),
/// positioned above the floating tab bar. Shared by shopping, meals, etc.
struct FloatingActionButton: View {
    let text: String
    var systemImage = "plus"
    let action: () -> Void

    var body: some View {
        GlassFAB(label: text, systemImage: systemImage, action: action)
            .padding(.trailing, 16)
            .padding(.bottom, 24)
            .accessibilityLabel(text)
    }
}

/// Icon-key picker grid presented in a sheet — used by shopping, meals, wishlists.
struct IconPickerSheet: View {
    let title: String
    let options: [String]
    let selected: String
    let symbolFor: (String) -> String
    let onPick: (String) -> Void
    /// When `onColorPick` is provided, a colour picker is shown too and applied live.
    var initialColor: Int?
    var onColorPick: ((Int?) -> Void)?

    @Environment(\.dismiss) private var dismiss
    @State private var color: Int?

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.lg) {
            ZStack {
                Text(title)
                    .font(.pushedTitle)
                    .foregroundStyle(Color.appOnSurface)
                HStack {
                    Button("Cancel") { dismiss() }
                        .font(.system(size: 17))
                        .foregroundStyle(Color.appPrimary)
                    Spacer()
                }
            }
            .frame(maxWidth: .infinity)
            IconGrid(options: options, selected: selected, symbolFor: symbolFor, onPick: onPick)
            if onColorPick != nil {
                SectionHeader(text: L("Color"))
                EventColorPicker(selection: Binding(
                    get: { color },
                    set: { color = $0; onColorPick?($0) }
                ))
            }
        }
        .padding(.horizontal, Spacing.screenEdge)
        .padding(.top, Spacing.lg)
        .padding(.bottom, Spacing.xl)
        .huggingSheet()
        .onAppear { color = initialColor }
    }
}

/// New-list sheet with name + icon picker — mirrors NewListDialog.
private struct NewListSheet: View {
    let onCreate: (String, String, Int?) -> Void

    @Environment(\.dismiss) private var dismiss
    @State private var title = ""
    @State private var selectedIcon = "shopping_cart"
    @State private var color: Int? = calendarEventColorPalette.first

    private var canCreate: Bool {
        !title.trimmingCharacters(in: .whitespaces).isEmpty
    }

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.lg) {
            SheetHeader(title: L("New list"), confirmTitle: L("Create"), confirmEnabled: canCreate) {
                dismiss()
            } onConfirm: {
                onCreate(title.trimmingCharacters(in: .whitespaces), selectedIcon, color)
                dismiss()
            }
            GlassField(systemImage: "cart", placeholder: L("List name"), text: $title)
            SectionHeader(text: L("Icon"))
            IconGrid(
                options: IconOptions.shopping,
                selected: selectedIcon,
                symbolFor: IconKeyMap.shoppingSymbol
            ) { selectedIcon = $0 }
            SectionHeader(text: L("Color"))
            EventColorPicker(selection: $color)
        }
        .padding(.horizontal, Spacing.screenEdge)
        .padding(.top, Spacing.lg)
        .padding(.bottom, Spacing.xl)
        .huggingSheet()
    }
}
